source/symbols/TypeHelpers.h

```diff
@@ -0,0 +1,208 @@
+     typedef struct {byte a[$]; bit b;} dest_t;
+     int a;
+     dest_t b = dest_t'(a);
+     */
+    auto width = type.getBitWidth();
+    if (width > 0)
+        return { 0, width };
+    if (type.isString())
+        return { destination == 1 ? 0 : CHAR_BIT, 0 };
+    bitwidth_t gcd = 0;
+    // TODO: check for overflow
+    if (type.isUnpackedArray()) {
+        auto [gcd1, width1] = linearCoefficients(*type.getArrayElementType(), destination);
+        const auto& ct = type.getCanonicalType();
+        if (ct.kind == SymbolKind::FixedSizeUnpackedArrayType) {
+            auto rw = ct.as<FixedSizeUnpackedArrayType>().range.width();
+            gcd = gcd1 * rw;
+            width = width1 * rw;
+        }
+        else if (destination) {
+            if (destination > 1) {
+                if (!gcd1)
+                    gcd = width1;
+                else {
+                    gcd = gcd1;
+                    width = width1;
+                }
+            }
+        }
+        else
+            gcd = std::gcd(gcd1, width1);
+    }
+    else if (type.isUnpackedStruct()) {
+        auto& us = type.getCanonicalType().as<UnpackedStructType>();
+        for (auto& field : us.membersOfType<FieldSymbol>()) {
+            auto [gcd1, width1] = linearCoefficients(field.getType(), destination);
+            if (destination > 1 && gcd1 > 0)
+                destination--;
+            gcd = std::gcd(gcd, gcd1);
+            width += width1;
+        }
+    }
+    // TODO: classes
+    return { gcd, width };
+}
+
+static ConstantValue constContainer(const Type& type, span<const ConstantValue> elems) {
+    switch (type.kind) {
+        case SymbolKind::FixedSizeUnpackedArrayType:
+        case SymbolKind::DynamicArrayType:
+        case SymbolKind::UnpackedStructType:
+            return ConstantValue::Elements(elems.cbegin(), elems.cend());
+        case SymbolKind::QueueType:
+            return SVQueue(elems.cbegin(), elems.cend());
+        default:
+            return nullptr;
+    }
+}
+
+using PackVector = decltype(std::declval<ConstantValue>().bitstream());
+static SVInt slicePacked(PackVector::const_iterator& iter, bitwidth_t& bit, bitwidth_t width) {
+    const ConstantValue* cp;
+    for (;;) {
+        cp = *iter;
+        if (cp->isInteger()) {
+            if (bit < cp->integer().getBitWidth())
+                break;
+        }
+        else {
+            if (bit < cp->str().length() * CHAR_BIT)
+                break;
+        }
+        bit = 0;
+        iter++;
+    }
+    if (cp->isInteger()) {
+        const auto& ci = cp->integer();
+        auto msb = ci.getBitWidth() - bit - 1;
+        auto lsb = std::min(bit + width, ci.getBitWidth());
+        lsb = ci.getBitWidth() - lsb;
+        bit += msb - lsb + 1;
+        ASSERT(bit <= ci.getBitWidth());
+        return ci.slice(static_cast<int32_t>(msb), static_cast<int32_t>(lsb));
+    }
+    else {
+        std::string_view str = cp->str();
+        auto byte0 = bit / CHAR_BIT;
```

Comment:
Instead of byte0 / byte1, perhaps firstByte / lastByte?

Reply:
Yes, I will change it.